### PR TITLE
Implement `tls_certificate`

### PR DESCRIPTION
**Pull request by wjam**

### CODE REVIEW

File: fastly/resource_fastly_tls_certificate.go [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+		return err
+	}
+
+	return nil
+}
+
+func resourceTLSCertificateUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*FastlyClient).conn
+
+	// API mandates the `CertBlob` field, but Fastly service doesn't allow the certificate to be updated
```
Comment (bengesoff) [inline fastly/resource_fastly_tls_certificate.go:156]: What do you mean by this?

File: website/docs/r/tls_certificate.html.markdown [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+resource "fastly_tls_private_key" "key" {
+  key_pem = tls_private_key.key.private_key_pem
+  name    = "tf-demo"
+}
+
+resource "fastly_tls_certificate" "example" {
+  name = "tf-demo"
+  certificate_body = tls_self_signed_cert.cert.cert_pem
+  depends_on = [fastly_tls_private_key.key]
```
Comment (bengesoff) [inline website/docs/r/tls_certificate.html.markdown:47]:
```suggestion
  depends_on = [fastly_tls_private_key.key_pem]
```

File: website_src/docs/r/tls_certificate.html.markdown.tmpl [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+resource "fastly_tls_private_key" "key" {
+  key_pem = tls_private_key.key.private_key_pem
+  name    = "tf-demo"
+}
+
+resource "fastly_tls_certificate" "example" {
+  name = "tf-demo"
+  certificate_body = tls_self_signed_cert.cert.cert_pem
+  depends_on = [fastly_tls_private_key.key]
```
Comment (bengesoff) [inline website_src/docs/r/tls_certificate.html.markdown.tmpl:47]:
```suggestion
  depends_on = [fastly_tls_private_key.key_pem]
```

File: fastly/provider_test.go [hunk 4]
```diff
@@ -47,6 +47,11 @@ func generateKey() (string, error) {
 }
 
 func generateKeyAndCert(SANs ...string) (key, cert string, err error) {
+	key, cert, _, err = generateKeyAndMultipleCerts(SANs...)
+	return
+}
+
+func generateKeyAndMultipleCerts(SANs ...string) (key, cert, cert2 string, err error) {
```
Comment (bengesoff) [inline fastly/provider_test.go:54]: Just to check my understanding: this is to enable you to have two certs with the same key right?

**Review by bengesoff — COMMENTED** [review]
Nice once, the `ValidateFunc` was a nice touch too 👌🏼

**Review by trentrosenbaum — CHANGES_REQUESTED** [review]
Looks good and I will execute in more detail.  The changes cover just a line in the docs for private key/cert order.

File: fastly/resource_fastly_tls_certificate.go (hunk 1 above)
Comment (wjam) [reply fastly/resource_fastly_tls_certificate.go:156]: An old comment I added before finding the comment in the API docs

File: website/docs/r/tls_certificate.html.markdown (hunk 2 above)
Comment (wjam) [reply website/docs/r/tls_certificate.html.markdown:47]: The resource isn't called `key_pem`?

File: website_src/docs/r/tls_certificate.html.markdown.tmpl (hunk 3 above)
Comment (wjam) [reply website_src/docs/r/tls_certificate.html.markdown.tmpl:47]: The resource isn't called `key_pem`?

File: fastly/provider_test.go (hunk 4 above)
Comment (wjam) [reply fastly/provider_test.go:54]: Yep

File: website/docs/r/tls_certificate.html.markdown (hunk 2 above)
Comment (bengesoff) [reply website/docs/r/tls_certificate.html.markdown:47]: Sorry read it wrong 🤦🏼

File: fastly/resource_fastly_tls_certificate.go [hunk 5]
```diff
@@ -0,0 +1,187 @@
+package fastly
```
Comment (bengesoff) [inline fastly/resource_fastly_tls_certificate.go:1]: Also I've been naming the TLS resources without `fastly` in the name - should I change those to be more in line with this naming or do you want to drop the `fastly` too?

Comment (bengesoff) [reply fastly/resource_fastly_tls_certificate.go:1]: The filename I mean

Comment (wjam) [reply fastly/resource_fastly_tls_certificate.go:1]: The pattern is to include the full name of the resource/data source in the file name, such as `aws_api_gateway_resource` and https://github.com/hashicorp/terraform-provider-aws/blob/master/aws/resource_aws_api_gateway_resource.go

Comment (bengesoff) [reply fastly/resource_fastly_tls_certificate.go:1]: No that's fine, just wanted to check so I can update mine 🙂 should that also apply to the function names then?

Comment (wjam) [reply fastly/resource_fastly_tls_certificate.go:1]: Yes, but this repository already contains `data_source_ip_ranges.go` - which lacks the provider in the file name but has a function in it which does include the provider name - and `resource_fastly_user_v1.go` - which includes the provider in the file name but doesn't include the provider in the function name.

Comment (bengesoff) [reply fastly/resource_fastly_tls_certificate.go:1]: Maybe something for TLC then 😂 ok no worries

**Review by bengesoff — APPROVED** [review]

**Review by trentrosenbaum — APPROVED** [review]

**Pull request merged**